File: labs-opensource/M09-rag/starter/rag_pipeline.py
"""
M09 Lab: The Full RAG Pipeline
===============================
Load → chunk → store → retrieve → generate-with-citations.
Run: python rag_pipeline.py   (from this folder; reads ../docs)
Requires: pip install openai "chromadb>=0.4.0"
"""

import glob
import os

import chromadb
from openai import OpenAI

DOCS_DIR = os.path.join(os.path.dirname(__file__), "..", "docs")


# ── Document Loading (COMPLETE) ──────────────────────────────
def load_documents(docs_dir: str) -> list[dict]:
    """Load all .md and .txt files from a directory."""
    docs = []
    for pattern in ["*.md", "*.txt"]:
        for path in glob.glob(os.path.join(docs_dir, pattern)):
            try:
                with open(path, "r", encoding="utf-8") as f:
                    docs.append({"content": f.read(), "source": os.path.basename(path)})
            except (IOError, UnicodeDecodeError) as e:
                print(f"  Skipping {path}: {e}")
    if not docs:
        raise FileNotFoundError(f"No documents found in {docs_dir}")
    return docs


# ── Part 1: Chunking (YOUR JOB) ──────────────────────────────
def chunk_text(text: str, chunk_size: int = 500, overlap: int = 50) -> list[str]:
    """Split text into overlapping chunks, cutting at natural boundaries.

    TODO:
    1. If len(text) <= chunk_size: return [text]
    2. Walk the text with a start pointer:
       a. end = start + chunk_size; chunk = text[start:end]
       b. If end < len(text): try separators ["\\n\\n", "\\n", ". "] in order —
          find the LAST occurrence (rfind) of the separator inside chunk;
          if it sits past the halfway point (last_sep > chunk_size * 0.5),
          cut there instead: end = start + last_sep + len(sep), re-slice chunk,
          and stop trying further separators
       c. Append chunk.strip()
       d. start = end - overlap     ← overlap for continuity
    3. Return chunks, dropping any empty strings
    """
    pass  # Remove this line when you add your code


# ── Part 2: Ingestion (COMPLETE) ─────────────────────────────
def ingest(docs_dir: str = DOCS_DIR) -> chromadb.Collection:
    """Load docs, chunk them, and store in ChromaDB."""
    print("-- Ingestion Pipeline --")
    docs = load_documents(docs_dir)
    print(f"  Loaded {len(docs)} documents")

    all_chunks = []
    for doc in docs:
        for i, chunk in enumerate(chunk_text(doc["content"], chunk_size=500, overlap=50)):
            all_chunks.append({"text": chunk, "source": doc["source"], "index": i})
    print(f"  Created {len(all_chunks)} chunks")

    # ChromaDB embeds with its built-in model — no embedding API needed
    client = chromadb.Client()  # in-memory; PersistentClient(path=...) for disk
    collection = client.get_or_create_collection(
        name="rag_lab", metadata={"hnsw:space": "cosine"}
    )
    collection.add(
        ids=[f"chunk_{i}" for i in range(len(all_chunks))],
        documents=[c["text"] for c in all_chunks],
        metadatas=[{"source": c["source"], "index": c["index"]} for c in all_chunks],
    )
    print(f"  Stored {collection.count()} chunks in ChromaDB")
    return collection


# ── Part 3: Query (YOUR JOB) ─────────────────────────────────
GROUNDING_SYSTEM = (
    "You are a helpful assistant that answers questions based ONLY on the "
    "provided context. If the context doesn't contain the answer, say "
    "\"I don't have enough information to answer that.\" Always cite your "
    "sources using [Source N] format."
)


def query_rag(collection, question: str, top_k: int = 3, verbose: bool = True) -> str:
    """Retrieve relevant chunks and generate a grounded, cited answer.

    TODO:
    1. results = collection.query(query_texts=[question], n_results=top_k)
       (try/except → return f"Retrieval error: {e}")
    2. If not results["documents"] or not results["documents"][0]:
         return "No relevant documents found. I don't have enough information."
    3. chunks = results["documents"][0]; sources = results["metadatas"][0]
       If verbose: print each retrieved source + first 60 chars
    4. Build the context block:
         f"[Source {i+1}: {meta['source']}]\\n{chunk}"  joined by "\\n\\n---\\n\\n"
    5. Call Mistral with GROUNDING_SYSTEM as the system message and a user
       message: f"Context:\\n{context}\\n\\nQuestion: {question}\\n\\n
                  Answer based on the context above, citing sources:"
       (try/except → return f"Generation error: {e}")
    6. Return the answer text
    """
    pass  # Remove this line when you add your code


# ── Test harness (COMPLETE) ──────────────────────────────────
if __name__ == "__main__":
    collection = ingest()

    questions = [
        "What happens if a continuation statement is filed late?",
        "What are the high-risk indicators in a lien risk assessment?",
        "What is the capital of France?",  # NOT in the docs — must refuse!
    ]
    for q in questions:
        print(f"\n{'=' * 60}\nQ: {q}")
        print(f"A: {query_rag(collection, q)}")

File: labs-opensource/M09-rag/starter/rag_pipeline.js
/**
 * M09 Lab: The Full RAG Pipeline (Node.js)
 * =========================================
 * Load → chunk → store → retrieve → generate-with-citations.
 *
 * NOTE: the chromadb JS client needs a Chroma SERVER running:
 *   pip install chromadb && chroma run --path ./chroma_data
 * The Python version runs fully in-process — prefer it if you only do one.
 *
 * Run: node rag_pipeline.js   (from this folder; reads ../docs)
 */

import { readFileSync, readdirSync } from "node:fs";
import { join, basename, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import OpenAI from "openai";
import { ChromaClient } from "chromadb";

const DOCS_DIR = join(dirname(fileURLToPath(import.meta.url)), "..", "docs");

// ── Document Loading (COMPLETE) ──────────────────────────────
function loadDocuments(docsDir) {
  const files = readdirSync(docsDir).filter((f) => f.endsWith(".md") || f.endsWith(".txt"));
  if (files.length === 0) throw new Error(`No documents found in ${docsDir}`);
  return files
    .map((f) => {
      try {
        return { content: readFileSync(join(docsDir, f), "utf-8"), source: basename(f) };
      } catch (e) {
        console.warn(`  Skipping ${f}: ${e.message}`);
        return null;
      }
    })
    .filter(Boolean);
}

// ── Part 1: Chunking (YOUR JOB) ──────────────────────────────
/**
 * Split text into overlapping chunks, cutting at natural boundaries.
 *
 * TODO:
 * 1. If text.length <= chunkSize: return [text]
 * 2. Walk the text with a start pointer:
 *    a. let end = start + chunkSize; let chunk = text.slice(start, end);
 *    b. If end < text.length: try separators ["\n\n", "\n", ". "] in order —
 *       find the LAST occurrence (lastIndexOf) inside chunk; if it sits past
 *       the halfway point (lastSep > chunkSize * 0.5), cut there instead:
 *       end = start + lastSep + sep.length, re-slice chunk, stop trying
 *    c. Push chunk.trim()
 *    d. start = end - overlap     ← overlap for continuity
 * 3. Return chunks, dropping empty strings
 */
function chunkText(text, chunkSize = 500, overlap = 50) {
  // TODO: implement
}

// ── Part 2: Ingestion (COMPLETE) ─────────────────────────────
async function ingest(docsDir = DOCS_DIR) {
  console.log("-- Ingestion Pipeline --");
  const docs = loadDocuments(docsDir);
  console.log(`  Loaded ${docs.length} documents`);

  const allChunks = [];
  for (const doc of docs) {
    chunkText(doc.content, 500, 50).forEach((chunk, i) => {
      allChunks.push({ text: chunk, source: doc.source, index: i });
    });
  }
  console.log(`  Created ${allChunks.length} chunks`);

  const chroma = new ChromaClient(); // expects chroma server on :8000
  const collection = await chroma.getOrCreateCollection({
    name: "rag_lab",
    metadata: { "hnsw:space": "cosine" },
  });
  await collection.add({
    ids: allChunks.map((_, i) => `chunk_${i}`),
    documents: allChunks.map((c) => c.text),
    metadatas: allChunks.map((c) => ({ source: c.source, index: c.index })),
  });
  console.log(`  Stored ${await collection.count()} chunks in ChromaDB`);
  return collection;
}

// ── Part 3: Query (YOUR JOB) ─────────────────────────────────
const GROUNDING_SYSTEM =
  "You are a helpful assistant that answers questions based ONLY on the " +
  "provided context. If the context doesn't contain the answer, say " +
  '"I don\'t have enough information to answer that." Always cite your ' +
  "sources using [Source N] format.";

/**
 * Retrieve relevant chunks and generate a grounded, cited answer.
 *
 * TODO:
 * 1. results = await collection.query({ queryTexts: [question], nResults: topK })
 *    (try/catch → return `Retrieval error: ${e.message}`)
 * 2. If (!results.documents?.[0]?.length):
 *      return "No relevant documents found. I don't have enough information.";
 * 3. const chunks = results.documents[0]; const sources = results.metadatas[0];
 * 4. Build the context block:
 *      `[Source ${i + 1}: ${sources[i].source}]\n${chunk}` joined by "\n\n---\n\n"
 * 5. Call Mistral with GROUNDING_SYSTEM as system message and the
 *    Context/Question user message (try/catch → `Generation error: ...`)
 * 6. Return the answer text
 */
async function queryRag(collection, question, topK = 3) {
  // TODO: implement
}

// ── Test harness (COMPLETE) ──────────────────────────────────
const collection = await ingest();

const questions = [
  "What happens if a continuation statement is filed late?",
  "What are the high-risk indicators in a lien risk assessment?",
  "What is the capital of France?", // NOT in the docs — must refuse!
];
for (const q of questions) {
  console.log(`\n${"=".repeat(60)}\nQ: ${q}`);
  console.log(`A: ${await queryRag(collection, q)}`);
}
